cpp/include/cudf/utilities/span.hpp

```diff
@@ -171,6 +181,86 @@ struct device_span : public span_base<T, Extent, device_span<T, Extent>> {
   constexpr device_span(C const& in) : base(thrust::raw_pointer_cast(in.data()), in.size())
   {
   }
+
+  template <typename OtherT,
+            std::size_t OtherExtent,
+            typename std::enable_if<(Extent == OtherExtent || Extent == dynamic_extent) &&
+                                      std::is_convertible<OtherT (*)[], T (*)[]>::value,
+                                    void>::type* = nullptr>
+  constexpr device_span(const device_span<OtherT, OtherExtent>& other) noexcept
+    : base(other.data(), other.size())
+  {
+  }
+};
+
+/**
+ * @brief Base class for 2D spans. Not compliant with STL container semantics/syntax.
+ *
+ * The index operator returns the corresponding row as a span.
+ */
+template <typename T>
+class base_2dspan {
+ public:
+  using size_type = std::pair<size_t, size_t>;
+
+  constexpr base_2dspan() noexcept = default;
+  constexpr base_2dspan(T* data, size_t rows, size_t columns) noexcept
+    : _data{data}, _size{rows, columns}
+  {
+  }
+  base_2dspan(T* data, size_type size) noexcept : _data{data}, _size{size} {}
+
+  constexpr auto data() const noexcept { return _data; }
+  constexpr auto size() const noexcept { return _size; }
+  constexpr auto count() const noexcept { return size().first * size().second; }
+  constexpr bool is_empty() const noexcept { return count() == 0; }
+
+  static constexpr size_t flatten_index(size_t row, size_t column, size_type size) noexcept
+  {
+    return row * size.second + column;
+  }
+
+ protected:
+  T* _data = nullptr;
+  size_type _size{0, 0};
+};
+
+template <typename T>
+class host_2dspan : public base_2dspan<T> {
```

Comment:
It looks like the redundancy between `host_2dspan` and `device_2dspan` could be reduced by templating `base_2dspan` on the `1Dspan_type`:

```
template <typename T, template <typename> typename 1DSpan>
class base_2dspan {
constexpr 1DSpan<T> operator[](size_t row)
  {
    return {this->data() + flatten_index(row, 0, this->size()), this->size().second};
  }

};

using host_2dspan = base_2dspan<T, host_span>;
using device_2dspan = base_2dspan<T, device_span>;
```

Reply:
Implemented this, with one difference: did not use template template parameter for 1DSpan, since the element type should match the element type of the 2D span.